usr.bin/rwho/rwho.1
.\"	$NetBSD: rwho.1,v 1.13 2005/10/01 00:03:47 wiz Exp $
.\"
.\" Copyright (c) 1983, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)rwho.1	8.1 (Berkeley) 6/6/93
.\"	$NetBSD: rwho.1,v 1.13 2005/10/01 00:03:47 wiz Exp $
.\"
.Dd September 30, 2005
.Dt RWHO 1
.Os
.Sh NAME
.Nm rwho
.Nd who is logged in on local machines
.Sh SYNOPSIS
.Nm
.Op Fl aHq
.Sh DESCRIPTION
The
.Nm
command produces output similar to
.Xr who 1 ,
but for all machines on the local network.
If no report has been
received from a machine for 11 minutes then
.Nm
assumes the machine is down, and does not report the users last known
to be logged into that machine.
.Pp
If a user hasn't typed to the system for a minute or more, then
.Nm
reports this idle time.
.Bl -tag -width Fl
.It Fl a
Include all users.
By default, if a user hasn't typed to the system for an hour or
more, then the user will be omitted from the output.
.It Fl H
Write column headings above the regular output.
.It Fl q
.Dq Quick mode :
List only the names and the number of users currently logged on.
When this option is used, all other options are ignored.
.El
.Sh FILES
.Bl -tag -width /var/rwho/whod.* -compact
.It Pa /var/rwho/whod.*
information about other machines
.El
.Sh SEE ALSO
.Xr finger 1 ,
.Xr rup 1 ,
.Xr ruptime 1 ,
.Xr rusers 1 ,
.Xr who 1 ,
.Xr rwhod 8
.Sh HISTORY
The
.Nm
command
appeared in
.Bx 4.3 .
.Sh BUGS
This is unwieldy when the number of machines
on the local net is large.


doc/CHANGES
LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.529 $>


[Note: This file does not mention every change made to the NetBSD source tree.
For an exhaustive list, please subscribe to the `source-changes' mailing list,
or see the mailing list archives.  For more information, send a message
containing just the word `help' to `majordomo@NetBSD.org'.]


Changes from NetBSD 3.0 to NetBSD 4.0:
	agr(4): Add pseudo-device driver for link aggregation. [yamt 20050318]
	kernel: Limit the number of locks per user to 1024 [christos 20050320]
	crypto: Upgrade to OpenSSL 0.9.7f [christos 20050325]
	attimer(4): Split the pcppi(4) driver to properly assign resources and
		allow both the devices to attach through ACPI [cube 20050325]
	gre(4): add support for IPv6 as inner protocol, contributed by
		Gert Doering [is 20050330]
	/rescue: Add tetris.  [lukem 20050401]
	uvm(9): Simplify kernel memory allocation API and its implementation.
		[yamt 20050401]
	ipf(8): Updated to version 4.1.8. [martti 20050403]
	identd(8): Add support for forwarding ident queries and receiving of
		proxied ident queries. [peter 20050404]
	ac97: Add support for AC'97 modems. [jmcneill 20050407]
	auich(4): Add support for Intel ICH AC'97 modems as audio devices.
		Currently disabled by default. [jmcneill 20050407]
	event(3): Updated libevent to 1.0c [provos 20050417]
	arc: Add a standalone boot program to load ELF kernels.
		[tsutsui 20050418]
	cvs(1): Upgrade to 1.11.20. [christos 20050419]
	openssh: Upgrade to 4.0. [christos 20050423]
	crypto: Upgrade to OpenSSL 0.9.7g [christos 20050423]
	gettext: Upgrade to 0.14.4 [christos 20050429]
	sparc64: Switch to wscons [martin 20050429]
	ehci(4): Various bug fixes.  The most visible is that low/full speed
		devices can now be connected to high speed hubs.
		[augustss 20050501]
	acpi(4): Import ACPI-CA 20050408 [kochi 20050502]
	psh3tp(4): Support touch-panel for PERSONA SH3 machines.
		   [kiyohara 20050523]
	evbarm: Add support for the Arcom Viper.  [pooka 20050606]
	scan_ffs(8): Added scan_ffs from OpenBSD. Utility to find FFSv1/FFSv2
		partitions, to recover lost disklabels. [xtraeme 20050615]
	ataraid(4): Add man page and adapt support for Adaptec HostRAID
		from FreeBSD. [briggs 20050620]
	wtf(6): Now searches all acronyms databases (the -t switch is
		gone), whatis(1) and then pkg_info(1). [hubertf 20050622]
	i386: Add cd9660 file system support to the BIOS bootloader.
		[junyoung 20050622]
	tlp(4): added support for ASIX AX88140A and AX88141.
		[rpaulo 20050624]
	pss(4): Remove the pss(4) driver.  It has never been workable.
		[kent 20050628]
	deroff(1): brought back from the dead. [perry 20050629]
	spell(1): brought back from the dead. [perry 20050629]
	ching(6):  brought back from the dead. [perry 20050630]
	pf(4): PF from OpenBSD 3.7 imported [peter 20050701]
	i386: WSDISPLAY_SCROLLSUPPORT is now enabled in GENERIC kernel.
		[rpaulo 20050702]
	awk(1): Upgrade to nawk 20050424 [jdolecek 20050702]
	ieee1394: ieee1394 from FreeBSD imported [kiyohara 20050711]
	ral(4): Add driver for Ralink PCI/Cardbus/USB wlan adapters
		(written by Damien Bergamini, ported by FUKAUMI Naoki)
		[drochner 20050714]
	vnd(4): Add support for configuring cloop2-compressed images
		read-only, written by Cliff Wright. [hubertf 20050716]
	uipaq(4): USB driver for iPAQ devices.  From OpenBSD.
		[augustss 20050718]
	netinet: Implement PMTU checks and ignore ICMP Source Quench messages
		for TCP as described in:
		http://www.gont.com.ar/drafts/icmp-attacks-against-tcp.html
		[christos 20050719]
	config(1): Implement "no <dev> at <attachment>". [cube 20050724]
	ucycom(4): Add driver for Cypress microcontroller based serial
		devices. [skrll 20050731]
	hpcmips: Add teliosio(4) driver for the Sharp Telios LCD screen and
		Battery unit. [nakayama 20050731]
	scan_lfs(8): Program to find LFS partitions on disks [xtraeme 20050731]
	ftpd(8): Implement option "-D", for running ftpd(8) in standalone mode.
		[peter 20050804]
	bpf(4): Add two new sysctls to retrieve statistics and current
		peers, viewable with netstat(1). [rpaulo 20050805]
	postfix: Updated to 2.2.5. [rpaulo 20050818]
	kernel: 64 bit inode changes [christos 20050819]
	getent(1): Add support for "ethers" database. [ginsbach 20050824]
	ftpd(8): Add new ftpd.conf(5) option, recvbufsize, and make
		readsize option also apply to network reads for transfers to
		the server. [ginsbach 20050824]
	sysinst: Added Spanish language translation. [xtraeme 20050826]
	UFS: Add experimental support for file-backed extended attributes
		on UFS1 file systems, adapted from FreeBSD.  [thorpej 20050828]
	psh3lcd(4): Support LCD screen driver for PERSONA SH3 machines.
		   [kiyohara 20050906]
	xge(4): Add driver for the Neterion Xframe-I 10 gigabit ethernet
		controller [ragge 20050909]
	i386: WSDISPLAY_SCROLLSUPPORT enabled in GENERIC.MPACPI,
		GENERIC_LAPTOP and XEN0 kernels. [rpaulo 20050909]
	amd64: WSDISPLAY_SCROLLSUPPORT enabled in GENERIC kernel.
		[rpaulo 20050909]
	kernel: Add tmpfs file system. [jmmv 20050910]
	mount_tmpfs(8): Add utility to mount a tmpfs file system.
		[jmmv 20050910]
	kernel/libc: Re-factor compatibility code to hide compatibility
		info from the header files and to allow libc to compile
		without the compatibility hooks [christos 20050912]
	am-utils: Upgrade to 6.1.2.1 [rpaulo 20050920]
	pam_afslog(8): Add module to obtain AFS tokens from krb5
		credentials and create a PAG if necessary. [tsarna 20050921]
	kernel: Reset the AMD Geode SC1100 microcontroller. [dyoung 20050922]
	geodewdog(4): Add driver for AMD Geode SC1100 microcontroller's
		watchdog timer. [dyoung 20050922]
	kernel: Add the VFS hooks interface so that kernel subsystems
		other than VFS can execute code upon VFS operations.
		[jmmv 20050923]
	kernel: Re-factor NFS exports list code handling, making it simpler
		by abstracting all the details into the NFS generic code,
		using the previously mentioned VFS hooks interface to make
		it work. [jmmv 20050923]

usr.bin/rwho/rwho.c
/*	$NetBSD: rwho.c,v 1.16 2005/10/01 13:33:21 he Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1983, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
/*static char sccsid[] = "from: @(#)rwho.c	8.1 (Berkeley) 6/6/93";*/
__RCSID("$NetBSD: rwho.c,v 1.16 2005/10/01 13:33:21 he Exp $");
#endif /* not lint */

#include <sys/param.h>
#include <sys/file.h>

#include <protocols/rwhod.h>

#include <dirent.h>
#include <err.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <fcntl.h>
#include <unistd.h>
#include <utmp.h>
#include <sysexits.h>

static void usage(void) __attribute__((__noreturn__));
static int utmpcmp(const void *, const void *);

/* 
 * this macro should be shared with ruptime.
 */
#define	DOWN(w,now)	((now) - (w).wd_recvtime > 11 * 60)
#define WHDRSIZE	(sizeof(struct whod) - \
    sizeof (((struct whod *)NULL)->wd_we))

struct	my_utmp {
	char	myhost[MAXHOSTNAMELEN];
	int	myidle;
	struct	outmp myutmp;
};

int
main(int argc, char **argv)
{
	DIR *dirp;
	struct dirent *dp;
	struct whoent *we;
	struct whod wd;
	struct my_utmp *mp, *start;
	time_t now;
	int ch;
	size_t width, n, i, nhosts, nusers, namount;
	int aflg, qflg, hflg;

	setprogname(argv[0]);
	(void)setlocale(LC_TIME, "");
	aflg = nusers = nhosts = qflg = hflg = 0;
	namount = 40;
	start = NULL;
	now = 0;

	while ((ch = getopt(argc, argv, "aHq")) != -1)
		switch(ch) {
		case 'a':
			aflg = 1;
			break;
		case 'q':
			qflg = 1;
			break;
		case 'H':
			hflg = 1;
			break;
		default:
			usage();
		}

	if(optind != argc)
		usage();

	if (qflg) {
		aflg = 0;
		hflg = 0;
	}

	if (chdir(_PATH_RWHODIR) || (dirp = opendir(".")) == NULL)
		err(EXIT_FAILURE, "Cannot access `%s'", _PATH_RWHODIR);

	(void)time(&now);

	if ((start = malloc(sizeof(*start) * namount)) == NULL)
		err(EXIT_FAILURE, "malloc");

	while ((dp = readdir(dirp)) != NULL) {
		int f;
		ssize_t cc;

		if (dp->d_ino == 0 || strncmp(dp->d_name, "whod.", 5))
			continue;

		f = open(dp->d_name, O_RDONLY);
		if (f < 0)
			continue;
		cc = read(f, &wd, sizeof (wd));
		if (cc < WHDRSIZE) {
			(void)close(f);
			continue;
		}
		nhosts++;
		if (DOWN(wd, now)) {
			(void)close(f);
			continue;
		}
		cc -= WHDRSIZE;
		we = wd.wd_we;
		for (n = cc / sizeof (struct whoent); n > 0; n--) {
			if (aflg == 0 && we->we_idle >= 60*60) {
				we++;
				continue;
			}

			if (nusers == namount) {
				namount = namount + 40;
				if ((start = realloc(start,
				    sizeof(struct my_utmp) * namount)) == NULL)
					err(1, "realloc");
			}

			mp = start + nusers;
	
			mp->myutmp = we->we_utmp;
			mp->myidle = we->we_idle;
			(void)strcpy(mp->myhost, wd.wd_hostname);
			nusers++; we++; mp++;
		}
		(void)close(f);
	}

	if (nhosts == 0)
		errx(EX_OK, "No hosts in `%s'.", _PATH_RWHODIR);

	mp = start;
	qsort(start, nusers, sizeof(*start), utmpcmp);
	width = 0;
	for (i = 0; i < nusers; i++) {
		size_t j = strlen(mp->myhost) + 1 + strlen(mp->myutmp.out_line);
		if (j > width)
			width = j;
		mp++;
	}
	mp = start;
	if (hflg) {
		(void)printf("%-*.*s %-*s %-12s %-*s\n", UT_NAMESIZE,
		    UT_NAMESIZE, "USER", (int)width, "LINE", "WHEN",
		    (int)width, "IDLE");
	}
	for (i = 0; i < nusers; i++) {
		char buf[BUFSIZ], cbuf[80];
		time_t t;

		if (qflg) {
			(void)printf("%-*.*s\n", UT_NAMESIZE, UT_NAMESIZE,
			    mp->myutmp.out_name);
			mp++;
			continue;
		}
		t = mp->myutmp.out_time;
		(void)strftime(cbuf, sizeof(cbuf), "%c", localtime(&t));
		(void)snprintf(buf, sizeof(buf), "%s:%s", mp->myhost,
		    mp->myutmp.out_line);
		(void)printf("%-*.*s %-*s %.12s", UT_NAMESIZE, UT_NAMESIZE,
		    mp->myutmp.out_name, (int)width, buf, cbuf + 4);
		mp->myidle /= 60;
		if (mp->myidle) {
			if (aflg) {
				if (mp->myidle >= 100 * 60)
					mp->myidle = 100 * 60 - 1;
				if (mp->myidle >= 60)
					(void)printf(" %2d", mp->myidle / 60);
				else
					(void)printf("   ");
			} else
				(void)printf(" ");
			(void)printf(":%02d", mp->myidle % 60);
		}
		(void)printf("\n");
		mp++;
	}

	if (qflg)
		(void)printf("# users = %zd\n", nusers);

	return EX_OK;
}

static int
utmpcmp(const void *v1, const void *v2)
{
	const struct my_utmp *u1, *u2;
	int rc;

	u1 = v1;
	u2 = v2;
	rc = strncmp(u1->myutmp.out_name, u2->myutmp.out_name, 8);
	if (rc)
		return rc;
	rc = strcmp(u1->myhost, u2->myhost);
	if (rc)
		return rc;
	return strncmp(u1->myutmp.out_line, u2->myutmp.out_line, 8);
}

static void
usage(void)
{
	(void)fprintf(stderr, "Usage: %s [-aHq]\n", getprogname());
	exit(1);
}
